### PR TITLE
feat(data): add DataStore query pagination unit tests

### PR DESCRIPTION
**Pull request by david-mcafee**
#### Description of changes

Adding unit tests that reproduce https://github.com/aws-amplify/amplify-js/issues/12049. See comments for which tests are failing.

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/__tests__/indexeddb.test.ts [hunk 1]
```diff
@@ -65,6 +69,25 @@ describe('Indexed db storage test', () => {
 		);
 	});
 
+	afterEach(async () => {
```
Comment (david-mcafee) [inline packages/datastore/__tests__/indexeddb.test.ts:72]: I ran into a strange error when reproducing this issue. In the final test, where I create a single record with a `page` value greater than `0` (which should _not_ return any records), I was received 10+ records that were not being returned by any of the tests. I attempted to clear / reset DataStore at a more fundamental level in this test, but decided to timebox the effort and add this step as a temporary cleanup solution.

File: packages/datastore/__tests__/indexeddb.test.ts [hunk 2]
```diff
@@ -605,6 +628,419 @@ describe('Indexed db storage test', () => {
...
+			limit: 100,
+		});
+
+		// For debugging only:
+		// Make sure test cleanup isn't interferring with result:
+		// const queryAll = await DataStore.query(Person, Predicates.ALL);
+		// expect(queryAll.length).toEqual(4);
+
+		// Note: there are a ton of records here. Test cleanup isn't happening..
+		// debugger;
```
Comment (svidgen) [inline packages/datastore/__tests__/indexeddb.test.ts:746]: ?

**Review by svidgen — COMMENTED** [review]
Looks like a crufty comment.

Also, could/should these tests by common adapter tests instead?

File: packages/datastore/__tests__/indexeddb.test.ts [hunk 3]
```diff
@@ -605,6 +628,403 @@ describe('Indexed db storage test', () => {
...
+		await DataStore.save(p1);
+
+		const secondPage = await DataStore.query(Person, Predicates.ALL, {
+			sort: s => s.firstName(SortDirection.DESCENDING),
+			page: 1,
+			limit: 20,
+		});
+
+		// Assertion fails - record is returned:
+		expect(secondPage.length).toEqual(0);
```
Comment (david-mcafee) [inline packages/datastore/__tests__/indexeddb.test.ts:1005]: This test currently fails - once we've found a fix, this should pass.

File: packages/datastore/__tests__/indexeddb.test.ts [hunk 4]
```diff
@@ -605,6 +628,403 @@ describe('Indexed db storage test', () => {
...
+		await DataStore.save(p1);
+
+		const thirdPage = await DataStore.query(Person, Predicates.ALL, {
+			sort: s => s.firstName(SortDirection.DESCENDING),
+			page: 3,
+			limit: 20,
+		});
+
+		// Assertion fails - record is returned:
+		expect(thirdPage.length).toEqual(0);
```
Comment (david-mcafee) [inline packages/datastore/__tests__/indexeddb.test.ts:1025]: This test currently fails - once we've found a fix, this should pass.

**Pull request closed without merging**